### PR TITLE
Ampers: Alex

### PR DESCRIPTION
**Pull request by brownav**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, how did you try querying the API? | I poked around the API documentation, then tested out url paths postman then found specific attributes in terminal with httparty gem.
Describe your API Wrapper.  How did you decide on the methods you created? | The api wrapper is simple, with two methods for recipe index (list_recipes) and show page (show_recipe). I used two different types of queries for each, which were explained in the Edamam api documentation.  
Describe an edge case or failure case test you wrote for your API Wrapper. | For the list_recipes api wrapper method, one tests ensures that if the search parameter is invalid, output for recipes will be nil. 
Explain how VCR aids in testing an API. | Making api calls can be 'expensive' with regards to limitations on amount of calls made, VCR allows us to save the results of a call which can be queried in further tests, rather than repeatedly calling the api for fresh results per test. 
What is the Heroku URL of your deployed application? | https://api-recipedia.herokuapp.com/
Provide a link to the Trello board you used | https://trello.com/b/WEqQgNFA/recipie-api

### CODE REVIEW

File: test/test_helper.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+VCR.configure do |config|
+  config.cassette_library_dir = 'test/cassettes' # folder where casettes will be located
+  config.hook_into :webmock # tie into this other tool called webmock
+  config.default_cassette_options = {
+    :record => :new_episodes,    # record new data when we don't have it yet
+    :match_requests_on => [:method, :uri, :body] # The http method, URI and body of a request all need to match
+  }
+  # Don't leave our Slack token lying around in a cassette file.
+  config.filter_sensitive_data("<SLACK_TOKEN>") do
```
Comment (CheezItMan) [inline test/test_helper.rb:24]: **SLACK_TOKEN**.... **SLACK_TOKEN**... Someone forgot to hide their API keys in the cassettes!

File: config/routes.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  root 'recipes#root', as: 'home'
+
+  resources :recipes, only: [:index]
+
+  get 'recipes/show', to: 'recipes#show', as: 'recipe'
```
Comment (CheezItMan) [inline config/routes.rb:8]: If you want to make this route RESTful, it should have a path `/recipes/:uri`

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  <link href="https://fonts.googleapis.com/css?family=Slabo+27px" rel="stylesheet">
+  <%= stylesheet_link_tag "application" %>
+  <%= javascript_include_tag "application", 'data-turbolinks-track' => true %>
+  <%= csrf_meta_tags %>
+</head>
+
+
+<body>
+
+  <div class="flash">
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:17]: `section` would make more semantic sense.

File: app/views/recipes/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  </div>
+</header>
+
+<h3 id='results'> Results for <%= params[:search] %> ...</h3>
+
+<section class="index">
+  <% @recipes.each do |recipe| %>
+  <div class="index-container">
+    <div>
+      <%= image_tag(recipe.image) %>
```
Comment (CheezItMan) [inline app/views/recipes/index.html.erb:19]: Making the image clickable as a link would improve the user experience.

File: test/controllers/recipes_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
+require "test_helper"
+require 'pry'
+
+describe RecipesController do
+
+  describe 'index' do
+    it "should get index path" do
+      VCR.use_cassette("recipes") do
+        get recipes_path
```
Comment (CheezItMan) [inline test/controllers/recipes_controller_test.rb:9]: You should have this be:  `get recipes_path(search: "pasta")`

Then add a test for a search term which returns no results.

File: test/controllers/recipes_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+      end
+
+      must_respond_with :ok
+    end
+
+  end
+
+  describe 'show' do
```
Comment (CheezItMan) [inline test/controllers/recipes_controller_test.rb:17]: Missing tests for the `show` action.

Something like
```ruby
it "will show a valid recipe" do
  VCR.use_cassette("recipes") do
    get recipe_path("http%3A%2F%2Fwww.edamam.com%2Fontologies%2Fedamam.owl%23recipe_95e8e9135376ecf57440f594d2e0c530")

    must_respond_with :ok
  end
end
```

Then create a test with an invalid URI.

File: test/lib/edamam_api_wrapper_test.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    it "can retrieve a valid recipe" do
+      VCR.use_cassette("recipes") do
+        response = EdamamApiWrapper.show_recipe("http://www.edamam.com/ontologies/edamam.owl#recipe_7bf4a371c6884d809682a72808da7dc2")
+
+        response.id.must_equal "http://www.edamam.com/ontologies/edamam.owl#recipe_7bf4a371c6884d809682a72808da7dc2"
+        response.must_be_kind_of Recipe
+      end
+    end
+
+    it "returns the correct recipe" do
```
Comment (CheezItMan) [inline test/lib/edamam_api_wrapper_test.rb:36]: You also need a test with an invalid URI.

File: lib/edamam_api_wrapper.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    if response['hits']
+      response['hits'].each do |recipe|
+        recipes_list << Recipe.new(recipe['recipe']['uri'], recipe['recipe']['label'], recipe['recipe']['image'], recipe['recipe']['url'], recipe['recipe']['source'], recipe['recipe']['ingredients'], recipe['recipe']['dietLabels'])
+      end
+    end
+    return recipes_list
+  end
+
+  def self.show_recipe(id)
+    response = HTTParty.get(URI.encode("#{URL}search?r=#{id}&app_id=#{ID}&app_key=#{KEY}"))
```
Comment (CheezItMan) [inline lib/edamam_api_wrapper.rb:22]: You need to do some sort of check to see if you could retrieve a recipe.

**Comment by CheezItMan** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** | 
Git hygiene | Good number of commits and good commit messages
Comprehension questions | Check
**General** | 
Rails fundamentals (RESTful routing, use of named paths) | Your `show` route is not RESTful.  Take a look at my notes in your `routs.rb` file.  
Semantic HTML | You are using `div` elements where more semantic tags like `section`, `article`, `header` etc.  
Errors are reported to the user | Nothing reported for empty search results and it crashes when given an invalid URI for the show action.
API Wrapper to handle the API requests | Check
Controller testing | **MISSING** some serious tests, see my inline notes.
Lib testing | Check, but you're missing some negative tests
Search Functionality | Check
List Functionality | Check
Show individual item functionality (link to original recipe opens in new tab) | Check, although it doesn't open in a new tab (minor)
**Styling** | 
Responsive layout | Check
List View shows 10 items at a time/pagination | Check
The app is styled to create an attractive user interface | Check, simple and effective
**API Features** | 
The App attributes Edaman | Check
The VCR casettes do not contain the API key | **OOOPS** You are **not** hiding the API key.  Check my notes in `test_helper.rb`
**External Resources** | 
Link to Trello Board | Check
Link to deployed app on Heroku | Check
**Overall** | You hit **most** of the learning goals.  You've got some serious issues with testing and other more minor issues.  Check my inline notes and let me know if you have questions.
